addons/web_editor/models/field_html_history.py

```diff
@@ -0,0 +1,191 @@
+        diff_split = diff_data_split[0].split('|', 1)
+        diff_lines = diff_split[1].split(',', 1)
+
+        tag_prefix = diff_split[0]
+        line_id = int(diff_lines[0])
+        end_line_id = int(diff_lines[1]) if len(diff_lines) > 1 else line_id
+        line_index = line_id - 1
+
+        text_diff = diff_data_split[1:] if len(diff_data_split) > 1 else []
+        text_diff.reverse()
+
+        if end_line_id > line_id:
+            for lid in range(end_line_id, line_id, -1):
+                if tag_prefix in ['-', 'R']:
+                    del current_content_list[lid - 1]
+
+        if tag_prefix in ['+', 'R']:
+            for line_text in text_diff:
+                current_content_list.insert(line_id, line_text)
+        if tag_prefix in ['-', 'R']:
+            del current_content_list[line_index]
+
+    return '<'.join(current_content_list)
+
+# ------------------------------------------------------------
+# Custom difflib Diff methods
+# ------------------------------------------------------------
+
+
+def _format_range_context(start, stop):
+    beginning = start + 1  # lines start numbering with one
+    length = stop - start
+    if not length:
+        beginning -= 1  # empty ranges begin at line just before the range
+    if length <= 1:
+        return '{}'.format(beginning)
+    return '{},{}'.format(beginning, beginning + length - 1)
+
+
+def _custom_diff(new_content_str, old_content_str):
+    new_content_lines = new_content_str.split('<')
+    old_content_lines = old_content_str.split('<')
+
+    prefix = dict(insert='+', delete='-', replace='R', equal='=')
+    for group in SequenceMatcher(None, new_content_lines, old_content_lines, False).get_grouped_opcodes(0):
+        diff_lines = []
+        first, last = group[0], group[-1]
+        diff_string = '|'
+
+        diff_string += _format_range_context(first[1], last[2])
+        if any(tag in {'replace', 'delete'} for tag, _, _, _, _ in group):
+            for tag, i1, i2, _, _ in group:
+                if tag not in {'insert', 'equal', 'replace'}:
+                    diff_string = prefix[tag] + diff_string
+        if any(tag in {'replace', 'insert'} for tag, _, _, _, _ in group):
+            for tag, _, _, j1, j2 in group:
+                if tag not in {'delete', 'equal'}:
+                    diff_string = prefix[tag] + diff_string
+                    for line in old_content_lines[j1:j2]:
+                        diff_lines.append(line)
+        if diff_lines:
+            yield str(diff_string) + '<' + '<'.join(diff_lines)
+        else:
+            yield str(diff_string)
+
+
+class HtmlHistory(models.Model):
+    _name = "field.html.history"
+    _description = "Field html History"
+
+    history_diff_ids = fields.One2many(
+        "field.html.history.diff", "related_id", string="Body History",
+        copy=True)
+
+    def write(self, vals):
+        history_field_name = getattr(self, 'HTML_HISTORY_FIELD_NAME')
```

Comment:
Would define a method, like "def _get_html_history_fname", returning False by default, and do something if we have a value. That way we override a method (find it better than class attributes), and we don't have to use getattr.

Reply:
Good idea, will do. :+1: